pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -0,0 +1,628 @@
+import asyncio
+import functools
+import json
+import logging
+import os
+import re
+import sys
+import traceback
+from collections import OrderedDict
+from pathlib import Path
+from typing import Dict, Iterable, List, Optional, Tuple
+from urllib.parse import quote
+
+import aiohttp
+import click
+import semver
+from doozerlib import assembly
+from doozerlib.util import brew_arch_for_go_arch, brew_suffix_for_arch
+from pyartcd import constants, exectools, util
+from pyartcd.cli import cli, click_coroutine, pass_runtime
+from pyartcd.exceptions import VerificationError
+from pyartcd.runtime import Runtime
+from semver import VersionInfo
+from tenacity import (RetryError, before_sleep_log, retry,
+                      retry_if_exception_type, retry_if_result,
+                      stop_after_attempt, wait_fixed)
+
+
+class PromotePipeline:
+    DEST_RELEASE_IMAGE_REPO = constants.RELEASE_IMAGE_REPO
+
+    def __init__(self, runtime: Runtime, group: str, assembly: str, release_offset: Optional[int],
+                 arches: Iterable[str], skip_blocker_bug_check: bool = False, skip_attached_bug_check: bool = False,
+                 skip_image_list: bool = False, permit_overwrite: bool = False) -> None:
+        self.runtime = runtime
+        self.group = group
+        self.assembly = assembly
+        self.release_offset = release_offset
+        self.arches = list(arches)
+        self.skip_blocker_bug_check = skip_blocker_bug_check
+        self.skip_attached_bug_check = skip_attached_bug_check
+        self.skip_image_list = skip_image_list
+        self.permit_overwrite = permit_overwrite
+
+        self._logger = self.runtime.logger
+        self._slack_client = self.runtime.new_slack_client()
+        self._mail = self.runtime.new_mail_client()
+
+        self._working_dir = self.runtime.working_dir
+        self._doozer_working_dir = self._working_dir / "doozer-working"
+        self._doozer_env_vars = os.environ.copy()
+        self._doozer_env_vars["DOOZER_WORKING_DIR"] = str(self._doozer_working_dir)
+        self._doozer_lock = asyncio.Lock()
+        self._elliott_working_dir = self._working_dir / "elliott-working"
+        self._elliott_env_vars = os.environ.copy()
+        self._elliott_env_vars["ELLIOTT_WORKING_DIR"] = str(self._elliott_working_dir)
+        self._elliott_lock = asyncio.Lock()
+        self._ocp_build_data_url = self.runtime.config.get("build_config", {}).get("ocp_build_data_url")
+        if self._ocp_build_data_url:
+            self._elliott_env_vars["ELLIOTT_DATA_PATH"] = self._ocp_build_data_url
+            self._doozer_env_vars["DOOZER_DATA_PATH"] = self._ocp_build_data_url
+
+    async def run(self):
```

Comment:
This method is a tad long. Would love to see it split up in more functions, and `run()` is calling these functions in sequence. This would help maintaining an overview for the reader, as well as give entrypoints for testing.
Not a blocker for merging :)

Reply:
It becomes longer as I was adding more functionalities to this job :)